vowpalwabbit/cb_to_cb_adf.cc

```diff
@@ -0,0 +1,186 @@
+      base.predict(data.adf_data.ecs);
+    }
+  }
+  else
+  {
+    base.predict(data.adf_data.ecs);
+  }
+
+  if (data.explore_mode) { ec.pred.a_s = data.adf_data.ecs[0]->pred.a_s; }
+  else
+  {
+    // cb_adf => first action is a greedy action TODO: is this a contract?
+    ec.pred.multiclass = data.adf_data.ecs[0]->pred.a_s[0].action + 1;
+  }
+}
+
+void output_example(vw& all, bool explore_mode, example& ec, CB::label& ld)
+{
+  if (explore_mode) { CB_EXPLORE::generic_output_example(all, ec, ld); }
+  else
+  {
+    float loss = CB_ALGS::get_cost_estimate(ld, ec.pred.multiclass);
+    CB_ALGS::generic_output_example(all, loss, ec, ld);
+  }
+}
+
+void finish_example(vw& all, cb_to_cb_adf& c, example& ec)
+{
+  output_example(all, c.explore_mode, ec, ec.l.cb);
+  VW::finish_example(all, ec);
+}
+
+/*
+    Purpose: run before cb, cb_explore, cbify and cb_adf related reductions
+    This will 'translate' cb (non adf) input commands into their cb_adf counterparts
+    Except when:
+        - the model file loaded is from a version older or including 8.9.0
+        - user bypasses this translation step using '--cb_force_legacy'
+        - user specifies the cb_type to 'dm' (not implemented in adf)
+
+    Related files: cb_algs.cc, cb_explore.cc, cbify.cc
+*/
+VW::LEARNER::base_learner* cb_to_cb_adf_setup(options_i& options, vw& all)
+{
+  bool compat_old_cb = false;
+  bool force_legacy = false;
+  bool eval = false;
+  std::string type_string = "mtr";
+  uint32_t num_actions;
+  uint32_t cbx_num_actions;
+  uint32_t cbi_num_actions;
+
+  option_group_definition new_options("Contextual Bandit Options");
+  new_options
+      .add(make_option("cb_to_cbadf", num_actions).necessary().help("Maps cb_adf to cb. Disable with cb_force_legacy."))
+      .add(make_option("cb", num_actions).keep().help("Maps cb_adf to cb. Disable with cb_force_legacy."))
+      .add(make_option("cb_explore", cbx_num_actions)
+               .keep()
+               .help("Translate cb explore to cbexploreadf. Disable with cb_force_legacy."))
+      .add(make_option("cbify", cbi_num_actions)
+               .keep()
+               .help("Translate cbify to cbexploreadf. Disable with cb_force_legacy."))
+      .add(make_option("cb_type", type_string).keep().help("contextual bandit method to use in {}"))
+      .add(make_option("eval", eval).help("Evaluate a policy rather than optimizing."))
+      .add(make_option("cb_force_legacy", force_legacy).keep().help("Default to old cb implementation"));
+
+  options.add_parse_and_check_necessary(new_options);
+
+  // ANY model created with older version should default to --cb_force_legacy
+  if (all.model_file_ver != EMPTY_VERSION_FILE) compat_old_cb = !(all.model_file_ver >= VERSION_FILE_WITH_CB_TO_CBADF);
+
+  // not compatible with adf
+  if (options.was_supplied("cbify_reg")) compat_old_cb = true;
+
+  // dm not implemented in cb_adf
+  if (type_string == "dm") compat_old_cb = true;
```

Comment:
```suggestion
  compat_old_cb = type_string == "dm";
```

Reply:
they can't since if type_string != "dm" it would end up putting a false on compat_old_cb and that would revert the previous case in line 119.